src/Coordination/KeeperDispatcher.cpp

```diff
@@ -83,30 +84,40 @@ void KeeperDispatcher::requestThread()
                 {
                     current_batch.emplace_back(request);
 
-                    /// Waiting until previous append will be successful, or batch is big enough
-                    /// has_result == false && get_result_code == OK means that our request still not processed.
-                    /// Sometimes NuRaft set errorcode without setting result, so we check both here.
-                    while (prev_result && (!prev_result->has_result() && prev_result->get_result_code() == nuraft::cmd_result_code::OK) && current_batch.size() <= max_batch_size)
+                    const auto try_get_request = [&]
                     {
                         /// Trying to get batch requests as fast as possible
-                        if (requests_queue->tryPop(request, 1))
+                        if (requests_queue->tryPop(request))
                         {
                             CurrentMetrics::sub(CurrentMetrics::KeeperOutstandingRequets);
                             /// Don't append read request into batch, we have to process them separately
                             if (!coordination_settings->quorum_reads && request.request->isReadRequest())
-                            {
                                 has_read_request = true;
-                                break;
-                            }
                             else
-                            {
-
                                 current_batch.emplace_back(request);
-                            }
+
+                            return true;
                         }
 
-                        if (shutdown_called)
-                            break;
+                        return false;
+                    };
+
+                    /// If we have enough requests in queue, we will try to batch at least max_quick_batch_size of them.
+                    size_t max_quick_batch_size = coordination_settings->max_requests_quick_batch_size;
+                    while (!shutdown_called && !has_read_request && current_batch.size() < max_quick_batch_size && try_get_request())
```

Comment:
Logically, there is a caveat here, - if some requests are batched and then there are no incoming requests, we'll stay in this loop and hold the requests in `current_batch`.